cmd/promtool/unittest.go

```diff
@@ -0,0 +1,467 @@
+			// then we compare alerts with the Eval at `ts`.
+			t := alertEvalTimes[curr]
+
+			presentAlerts := alertsInTest[t]
+			got := make(map[string]labelsAndAnnotations)
+
+			// Same Alert name can be present in multiple groups.
+			// Hence we collect them all to check against expected alerts.
+			for _, g := range groups {
+				grules := g.Rules()
+				for _, r := range grules {
+					ar, ok := r.(*rules.AlertingRule)
+					if !ok {
+						continue
+					}
+					if _, ok := presentAlerts[ar.Name()]; !ok {
+						continue
+					}
+
+					var alerts labelsAndAnnotations
+					for _, a := range ar.ActiveAlerts() {
+						if a.State == rules.StateFiring {
+							alerts = append(alerts, labelAndAnnotation{
+								Labels:      append(labels.Labels{}, a.Labels...),
+								Annotations: append(labels.Labels{}, a.Annotations...),
+							})
+						}
+					}
+
+					got[ar.Name()] = append(got[ar.Name()], alerts...)
+				}
+			}
+
+			for _, testcase := range alertTests[t] {
+				// Checking alerts.
+				gotAlerts := got[testcase.Alertname]
+
+				var expAlerts labelsAndAnnotations
+				for _, a := range testcase.ExpAlerts {
+					// User gives only the labels from alerting rule, which doesn't
+					// include this label (added by Prometheus during Eval).
+					a.ExpLabels[labels.AlertName] = testcase.Alertname
+
+					expAlerts = append(expAlerts, labelAndAnnotation{
+						Labels:      labels.FromMap(a.ExpLabels),
+						Annotations: labels.FromMap(a.ExpAnnotations),
+					})
+				}
+
+				if gotAlerts.Len() != expAlerts.Len() {
+					errs = append(errs, fmt.Errorf("    alertname:%s, time:%s, \n        exp:%#v, \n        got:%#v",
+						testcase.Alertname, testcase.EvalTime.String(), expAlerts.String(), gotAlerts.String()))
+				} else {
+					sort.Sort(gotAlerts)
+					sort.Sort(expAlerts)
+
+					if !reflect.DeepEqual(expAlerts, gotAlerts) {
+						errs = append(errs, fmt.Errorf("    alertname:%s, time:%s, \n        exp:%#v, \n        got:%#v",
+							testcase.Alertname, testcase.EvalTime.String(), expAlerts.String(), gotAlerts.String()))
+					}
+				}
+			}
+
+			curr++
+		}
+	}
+
+	// Checking promql expressions.
+	for _, testCase := range tg.PromqlExprTests {
+		got, err := query(suite.Context(), testCase.Expr, mint.Add(testCase.EvalTime),
+			suite.QueryEngine(), suite.Queryable())
+		if err != nil {
+			errs = append(errs, fmt.Errorf("    expr:'%s', time:%s, err:%s", testCase.Expr,
+				testCase.EvalTime.String(), err.Error()))
+		}
+
+		var gotSamples []parsedSample
+		for _, s := range got {
+			gotSamples = append(gotSamples, parsedSample{
+				Labels: s.Metric.Copy(),
+				Value:  s.V,
+			})
+		}
+
+		var expSamples []parsedSample
+		parseError := false
+		for _, s := range testCase.ExpSamples {
+			lb, err := promql.ParseMetric(s.Labels)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("    expr:'%s', time:%s, err:%s", testCase.Expr,
+					testCase.EvalTime.String(), err.Error()))
+				parseError = true
```

Comment:
Why not break out here, rather than adding a new bool?

Reply:
So that I can skip checking for `got ==  exp` [here](https://github.com/codesome/prometheus/blob/5576d8e6312d07431d2db5e8d76ac78908510968/cmd/promtool/unittest.go#L301) in case it was an error. So I am using a bool to keep track of that.

Reply:
If you break out of the outer for, you don't need that

Reply:
I have made the changes to continue to the outer for.